runtime/slot-manager.js

```diff
@@ -11,81 +11,145 @@
 
 const assert = require('assert');
 
-// TODO(sjmiles): SlotManager
-// - needs a new name
-// - needs better function names
-// - needs better data design
+class Slot {
+  constructor(slotid) {
+    this._slotid = slotid;
+    this._dom = undefined;
+    this._exposedView = undefined;
+    this._particleSpec = undefined;
+    this._pendingRequestsHandlers = [];
+  }
+  get slotid() { return this._slotid; }
+  get particleSpec() { return this._particleSpec; }
+  associateWithParticle(particleSpec) {
+    assert(!this._particleSpec, "Particle spec already set, cannot associate slot");
+    assert(!this._exposedView ||
+           this._particleSpec.renderMap.get(this._slotid) == this._exposedView,
+           "Cannot associate particle-spec with an unmatching view.");
+
+    this._particleSpec = particleSpec;
+  }
+  disassociateParticle() {
+    assert(this._particleSpec, "Particle spec is not set, cannot disassociate slot");
+    this._particleSpec = undefined;
+  }
+  isAssociated() {
+    return !!this._particleSpec;
+  }
+  clearDom() {
+    this._dom = null;
+  }
+  initializeDom(dom, exposedView) {
+    this._dom = dom;
+    this.exposedView = exposedView;
+  }
+  isDomInitialized() {
+    return !!this._dom;
+  }
+  get domContent() {
+    return this._dom ? this._dom.innerHTML : undefined;
+  }
+  setDomContent(content) {
+    assert(!!this._dom, "Dom isn't initialized, cannot set content");
+    this._dom.innerHTML = content;
+  }
+  findInnerSlots() {
+    if (global.document) {
+      return Array.from(this._dom.querySelectorAll("[slotid]"));
+    } else {
+      var slots = [];
+      var slot;
+      var RE = /slotid="([^"]*)"/g;
+      while ((slot = RE.exec(this._dom.innerHTML))) {
+        slots.push({id:slot[1]});
+      }
+      return slots;
+    }
+  }
+  findEventGenerators() {
+    if (global.document) {
+      return this._dom.querySelectorAll('[events]');
+    }
+    // TODO(mmandlis): missing mock-DOM version
+    return [];
+  }
+
+  addPendingRequest(handler) {
+    this._pendingRequestsHandlers.push(handler);
+  }
+  providePendingSlot() {
+    assert(!this.isAssociated(), "Cannot provide associated slot.");
+    if (this._pendingRequestsHandlers.length > 0) {
+      this._pendingRequestsHandlers[0]();
+      this._pendingRequestsHandlers.splice(0, 1);
+    }
+  }
+}
 
 class SlotManager {
   constructor(domRoot, pec) {
-    this._content = {};
-    this._slotDom = {root: {insertion: domRoot, view: undefined}};
-    this._slotOwners = {};
-    this._targetSlots = new Map();
-    this._pendingSlotRequests = {};
+    this._slotsMap = new Map();
+    this._getOrCreateSlot('root').initializeDom(domRoot, /* exposedView= */ undefined);
+
+    this._reverseSlotMap = new Map();  // slotid by particle-spec.
     this._pec = pec;
   }
-  registerSlot(particleid, slotid, view) {
+  registerSlot(particleSpec, slotid) {
```

Comment:
No, it's not an ID, is it particle+views, this is why i renamed it, because ID was confusing. 
In outer PEC here:
https://github.com/PolymerLabs/arcs/blob/master/runtime/outer-PEC.js#L106
it is called particleSpec, so this is why I used this name.